ntoskrnl/io/pnpmgr/pnpmgr.c

```diff
@@ -1760,6 +1765,79 @@ IopGetParentIdPrefix(PDEVICE_NODE DeviceNode,
     return Status;
 }
 
+static
+ULONG
+IopValidateID(IN PWCHAR Id,
+              IN SIZE_T MaxIdLen,
+              IN ULONG MaxSeparators,
+              IN BOOLEAN IsMultiSz)
+{
+   PWCHAR PtrSymbol;
+   PWCHAR StringEnd;
+   WCHAR Symbol;
+   ULONG SeparatorsCount = 0;
+   PWCHAR PtrPrevSymbol = NULL;
+
+   PAGED_CODE();
+
+   StringEnd = Id + MaxIdLen;
+
+   for (PtrSymbol = Id; PtrSymbol < StringEnd; PtrSymbol++)
+   {
+      Symbol = *PtrSymbol;
+
+      if (Symbol == UNICODE_NULL)
+      {
+         if (!IsMultiSz || (PtrPrevSymbol && PtrSymbol == PtrPrevSymbol + 1))
+         {
+            if (MaxSeparators == SeparatorsCount ||
+                MaxSeparators == MAX_SEPARATORS_MULTISZ)
+            {
+               return (ULONG)(PtrSymbol - Id) + 1;
+            }
+
+            DPRINT("IopValidateID: SeparatorsCount - %lu, MaxSeparators - %lu\n",
+                   SeparatorsCount, MaxSeparators);
+
+            // FIXME logging
+            return 0;
+         }
+
+         StringEnd += MaxIdLen;
+         PtrPrevSymbol = PtrSymbol;
+      }
+      else if (Symbol < ' ' || Symbol > 0x7F || Symbol == ',')
+      {
+         DPRINT("IopValidateID: Invalid character - %04X\n", Symbol);
+
+         // FIXME logging
+         return 0;
+      }
+      else if (Symbol == ' ')
+      {
+         *PtrSymbol = '_';
+      }
+      else if (Symbol == '\\')
+      {
+         SeparatorsCount++;
+
+         if (SeparatorsCount > MaxSeparators)
```

Comment:
What happens if you're analyzing IDs in a multi-string, with each ID having a number of separators < MaxSeparators?

Reply:
Example:
```
ACPI\VEN_ACPI&DEV_000C
ACPI\ACPI000C
*ACPI000C
```

Reply:
For MultiSz: MaxSeparators == MAX_SEPARATORS_MULTISZ == 0xFFFFFFFF